### PR TITLE
polish(lxl&zms): add middleware comments

### PR DESCRIPTION
**Pull request by hiha3456**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/framework/middleware/collector.py [hunk 1]
```diff
@@ -11,6 +11,17 @@
 
 
 class StepCollector:
+    """
+    Overview:
+        The middleware that executes the collection by steps, including model inference and transition process.
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config.
+        - policy (:obj:`Policy`): The policy to be collected.
```
Comment (hiha3456) [inline ding/framework/middleware/collector.py:19]: Here : Policy is ok I think

File: ding/framework/middleware/functional/collector.py [hunk 2]
```diff
@@ -47,9 +47,27 @@ def clear(self):
...
+        - env (:obj:`BaseEnvManager`): The env where the inference process is performed.
+            The env.ready_obs (:obj:`tnp.array`) will be used as model input.
+    """
+
     env.seed(cfg.seed)
 
     def _inference(ctx: "OnlineRLContext"):
+        """
+        Output of ctx:
+            - obs (:obj:`Dict[Tenosr]`): The input states fed into the model.
```
Comment (hiha3456) [inline ding/framework/middleware/functional/collector.py:65]: obs (:obj:`Dict[Tensor]`)

File: ding/framework/middleware/functional/collector.py [hunk 3]
```diff
@@ -47,9 +47,27 @@ def clear(self):
...
+            - action: (:obj:`List[np.ndarray]`): The infered actions listed by env_id.
```
Comment (hiha3456) [inline ding/framework/middleware/functional/collector.py:66]: "inferred" actions

File: ding/framework/middleware/functional/evaluator.py [hunk 4]
```diff
@@ -146,10 +146,28 @@ def get_episode_info(self) -> dict:
...
+        - env (:obj:`BaseEnvManager`): The env for the evaluation.
+    """
+
     env.seed(cfg.seed, dynamic_seed=False)
 
     def _evaluate(ctx: "OnlineRLContext"):
-        # evaluation will be executed if the task begins or enough train_iter after last evaluation
+        """
+        Overview:
+            - The evaluation will be executed if the task begins and enough train_iter after last evaluation.
```
Comment (hiha3456) [inline ding/framework/middleware/functional/evaluator.py:163]: Maybe "enough train_iter passed"?

File: ding/framework/middleware/functional/trainer.py [hunk 5]
```diff
@@ -10,10 +10,26 @@
...
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config.
+        - policy (:obj:`Policy`): The policy to be trained in step-by-step mode.
+    """
 
     def _train(ctx: Union["OnlineRLContext", "OfflineRLContext"]):
+        """
+        Input of ctx:
+            - train_data (:obj:`Dict`): The data used to update the network. 
+                It will trains only if the data is not empty.
```
Comment (hiha3456) [inline ding/framework/middleware/functional/trainer.py:25]: It will train

File: ding/framework/middleware/ckpt_handler.py [hunk 6]
```diff
@@ -23,6 +35,17 @@ def __init__(self, cfg: EasyDict, policy: Policy, train_freq: Optional[int] = No
...
+        """
+        Overview:
+            The method to save checkpoint data. \
+            The checkpoint data will be saved in a file in following 3 cases: \
+                - Every time when train self.train_freq iterations; \
+                - Every time when the eval reward is the best eval reward so far; \
+                - When task.finish is True. \
+        Input of ctx:
+            - train_iter (:obj:`int`): Number of train_iter.
+            - eval_value (:obj:`int`): The eval reward of current iteration.
```
Comment (PaParaZz1) [inline ding/framework/middleware/ckpt_handler.py:47]: `eval_value` is not always int type, float can be more general.

File: ding/framework/middleware/ckpt_handler.py [hunk 7]
```diff
@@ -23,6 +35,17 @@ def __init__(self, cfg: EasyDict, policy: Policy, train_freq: Optional[int] = No
...
     def __call__(self, ctx: Union["OnlineRLContext", "OfflineRLContext"]) -> None:
+        """
+        Overview:
+            The method to save checkpoint data. \
+            The checkpoint data will be saved in a file in following 3 cases: \
+                - Every time when train self.train_freq iterations; \
+                - Every time when the eval reward is the best eval reward so far; \
+                - When task.finish is True. \
+        Input of ctx:
+            - train_iter (:obj:`int`): Number of train_iter.
```
Comment (PaParaZz1) [inline ding/framework/middleware/ckpt_handler.py:46]: Number of training iteration, i.e. the number of updating policy related network

File: ding/framework/middleware/collector.py [hunk 8]
```diff
@@ -42,6 +59,17 @@ def __call__(self, ctx: "OnlineRLContext") -> None:
 
 
 class EpisodeCollector:
+    """
```
Comment (PaParaZz1) [inline ding/framework/middleware/collector.py:62]: move to `__init__` method, class comment should contain the whole overview and interface method list

File: ding/framework/middleware/collector.py [hunk 9]
```diff
@@ -42,6 +59,17 @@ def __call__(self, ctx: "OnlineRLContext") -> None:
...
+    Overview:
+        The middleware that executes the collection by episodes, including model inference and transition process.
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config.
+        - policy (:obj:`Policy`): The policy to be collected.
+        - env (:obj:`BaseEnvManager`): The env for the collection.
```
Comment (PaParaZz1) [inline ding/framework/middleware/collector.py:68]: add proper newline notation

File: ding/framework/middleware/functional/advantage_estimator.py [hunk 10]
```diff
@@ -14,10 +14,31 @@
 
 
 def gae_estimator(cfg: EasyDict, policy: Policy, buffer_: Optional[Buffer] = None) -> Callable:
+    """
+    Overview:
+        Calculate value using observation of input data, then call function gae to get advantage. \
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/advantage_estimator.py:19]: call function `gae`

File: ding/framework/middleware/functional/explorer.py [hunk 11]
```diff
@@ -22,9 +36,17 @@ def _eps_greedy(ctx: "OnlineRLContext"):
 
 
 def eps_greedy_masker():
+    """
+    Overview:
+        The middleware that returns masked epsilon value.
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/explorer.py:41]: add reason about why masks eps

File: ding/framework/middleware/learner.py [hunk 12]
```diff
@@ -41,6 +55,15 @@ def __call__(self, ctx: "OnlineRLContext") -> None:
 
 
 class HERLearner:
+    """
+    Overview:
+        The middleware that executes the whole learning process with HER reward.
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config.
+        - policy (:obj:`Policy`): The policy to be trained.
+        - buffer_: The replay buffer to store the data for training.
```
Comment (PaParaZz1) [inline ding/framework/middleware/learner.py:64]: no corresponding type

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/framework/middleware/functional/collector.py [hunk 13]
```diff
@@ -66,11 +84,32 @@ def _inference(ctx: "OnlineRLContext"):
...
 def rolloutor(cfg: EasyDict, policy: Policy, env: BaseEnvManager, transitions: TransitionList) -> Callable:
+    """
+    Overview:
+        The middleware that executes the transition process in the env.
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config.
+        - policy (:obj:`Policy`): The policy to be used during transition.
+        - env (:obj:`BaseEnvManager`): The env where the transition process is performed.
+            BaseEnvManager object or its derivatives are supported.
+        - transitions (:obj:`TransitionList`): The transition information which will be filled in this process.
```
Comment (davide97l) [inline ding/framework/middleware/functional/collector.py:95]: Could be given more details about this transition information?

File: ding/framework/middleware/functional/data_processor.py [hunk 14]
```diff
@@ -35,8 +50,39 @@ def _push(ctx: "OnlineRLContext"):
...
+    Overview:
+        The return function is a generator which meanly fetch a batch of data from a buffer, \
+        a list of buffers, or a dict of buffers.
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config which should contain following keys: ['cfg.policy.learn.batch_size'].
+        - buffer_ (:obj:`Union[Buffer, List[Tuple[Buffer, float]], Dict[str, Buffer]]`): \
+            The buffer where the data is fetched from. \
+            ``Buffer`` type means a buffer.\
+            ``List[Tuple[Buffer, float]]`` type means a list of tuple. In each tuple there is a buffer and a float. \
+            The float defines that, how many times of batch_size the size of data is \
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:63]: remove 'that,'

Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:63]: how many `batch_size` is the size of the data

File: ding/framework/middleware/functional/data_processor.py [hunk 15]
```diff
@@ -35,8 +50,39 @@ def _push(ctx: "OnlineRLContext"):
...
+            which sampled from the corresponding buffer.\
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:64]: which is sampled

File: ding/framework/middleware/functional/data_processor.py [hunk 16]
```diff
@@ -35,8 +50,39 @@ def _push(ctx: "OnlineRLContext"):
...
+        """
+        Input of ctx:
+            - train_output (:obj:`Union[Dict, Deque[Dict]]`): This attribute should exist \
+                if buffer_ is of type Buffer and if buffer_ use the middleware PriorityExperienceReplay. \
+                The meta data ``priority`` of the sampled data in the buffer_ will be updated \
+                to the ``priority`` attribute of ctx.train_output if ctx.train_output is a dict, \
+                or the ``priority`` attribute of ctx.train_output's popped element \
+                if ctx.train_output is a deque of dicts.
+        Output of ctx:
+            - train_data (:obj:`Union[List[Dict], Dict[str, List[Dict]]]`): the fetched data. \
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:80]: The fetched

File: ding/framework/middleware/functional/data_processor.py [hunk 17]
```diff
@@ -92,10 +138,29 @@ def _fetch(ctx: "OnlineRLContext"):
 
 
 def offline_data_fetcher(cfg: EasyDict, dataset: Dataset) -> Callable:
+    """
+    Overview:
+        The outer function transform a Pytorch Dataset to DataLoader. \
+        And The return function is a generator which each time fetch a batch of data from the previous DataLoader.\
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:144]: remove 'And'

File: ding/framework/middleware/functional/data_processor.py [hunk 18]
```diff
@@ -92,10 +138,29 @@ def _fetch(ctx: "OnlineRLContext"):
 
 
 def offline_data_fetcher(cfg: EasyDict, dataset: Dataset) -> Callable:
+    """
+    Overview:
+        The outer function transform a Pytorch Dataset to DataLoader. \
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:143]: transforms

File: ding/framework/middleware/functional/data_processor.py (hunk 17 above)
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:144]: fetches

File: ding/framework/middleware/functional/data_processor.py [hunk 19]
```diff
@@ -92,10 +138,29 @@ def _fetch(ctx: "OnlineRLContext"):
...
+        And The return function is a generator which each time fetch a batch of data from the previous DataLoader.\
+        Please refer to the link https://pytorch.org/tutorials/beginner/basics/data_tutorial.html \
+        and https://pytorch.org/docs/stable/data.html for more details.
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config which should contain following keys: ['cfg.policy.learn.batch_size'].
+        - dataset (:obj:`Dataset`): The dataset of type torch.utils.data.Dataset which store the data.
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:149]: stores

File: ding/framework/middleware/functional/data_processor.py [hunk 20]
```diff
@@ -107,8 +172,22 @@ def _fetch(ctx: "OfflineRLContext"):
 
 
 def offline_data_saver(cfg: EasyDict, data_path: str, data_type: str = 'hdf5') -> Callable:
+    """
+    Overview:
+        Save the expert data of offline RL in a path.
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:177]: in a directory

File: ding/framework/middleware/functional/data_processor.py [hunk 21]
```diff
@@ -117,8 +196,21 @@ def _save(ctx: "OnlineRLContext"):
 
 
 def sqil_data_pusher(cfg: EasyDict, buffer_: Buffer, expert: bool) -> Callable:
+    """
+    Overview:
+        Push trajectories into buffer in sqil learning pipeline.
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:201]: the buffer

File: ding/framework/middleware/functional/data_processor.py [hunk 22]
```diff
@@ -117,8 +196,21 @@ def _save(ctx: "OnlineRLContext"):
...
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config.
+        - buffer_ (:obj:`Buffer`): Buffer to push the data in.
+        - expert (:obj:`bool`): Wether the pushed data is expert data or not. \
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:205]: Whether

File: ding/framework/middleware/functional/data_processor.py [hunk 23]
```diff
@@ -117,8 +196,21 @@ def _save(ctx: "OnlineRLContext"):
...
+            In each element of the pushed data, the reward will be set to one if this attribute is True, otherwise zero.
```
Comment (davide97l) [inline ding/framework/middleware/functional/data_processor.py:206]: 1 and 0

File: ding/framework/middleware/functional/enhancer.py [hunk 24]
```diff
@@ -10,16 +10,41 @@
 
 
 def reward_estimator(cfg: EasyDict, reward_model: "BaseRewardModel") -> Callable:
+    """
+    Overview:
+        Estimate reward of train_data using reward_model.
```
Comment (davide97l) [inline ding/framework/middleware/functional/enhancer.py:15]: the reward

File: ding/framework/middleware/functional/enhancer.py [hunk 25]
```diff
@@ -10,16 +10,41 @@
...
+        """
         reward_model.estimate(ctx.train_data)  # inplace modification
 
     return _enhance
 
 
 def her_data_enhancer(cfg: EasyDict, buffer_: "Buffer", her_reward_model: "HerRewardModel") -> Callable:
+    """
+    Overview:
+        Fetch a batch of data/a episode from buffer_, \
```
Comment (davide97l) [inline ding/framework/middleware/functional/enhancer.py:34]: data/episode

File: ding/framework/middleware/functional/enhancer.py [hunk 26]
```diff
@@ -10,16 +10,41 @@
...
+        then use her_reward_model to get HER processed episodes from original episodes.
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config which should contain following keys \
+            if her_reward_model.episode_size is None: ['cfg.policy.learn.batch_size'].
+        - buffer_ (:obj:`Buffer`): Buffer to sample data from.
+        - her_reward_model (:obj:`HerRewardModel`): Hindsight Experience Replay model which is used to process episodes.
```
Comment (davide97l) [inline ding/framework/middleware/functional/enhancer.py:40]: Hindsight Experience Replay (HER). So that later is clear that HER refers to it

File: ding/framework/middleware/functional/explorer.py [hunk 27]
```diff
@@ -7,10 +7,24 @@
 
 
 def eps_greedy_handler(cfg: EasyDict) -> Callable:
+    """
+    Overview:
+        The middleware that compute epsilon value according to the env_step.
```
Comment (davide97l) [inline ding/framework/middleware/functional/explorer.py:12]: computes

File: ding/framework/middleware/functional/trainer.py [hunk 28]
```diff
@@ -10,10 +10,26 @@
...
+    Arguments:
+        - cfg (:obj:`EasyDict`): Config.
+        - policy (:obj:`Policy`): The policy to be trained in step-by-step mode.
+    """
 
     def _train(ctx: Union["OnlineRLContext", "OfflineRLContext"]):
+        """
+        Input of ctx:
+            - train_data (:obj:`Dict`): The data used to update the network.
+                It will trains only if the data is not empty.
```
Comment (davide97l) [inline ding/framework/middleware/functional/trainer.py:25]: 'will train' or 'trains'

File: ding/framework/middleware/functional/trainer.py [hunk 29]
```diff
@@ -10,10 +10,26 @@
...
+            - train_iter: (:obj:`int`): The training iteration count.
+                The log will be printed if it reachs certain values.
```
Comment (davide97l) [inline ding/framework/middleware/functional/trainer.py:27]: once it reaches

**Review by PaParaZz1 — APPROVED** [review]
